avro-fastserde/src/main/java/com/linkedin/avro/fastserde/FastSerdeCache.java

```diff
@@ -0,0 +1,529 @@
+              if (Supplier.class.isAssignableFrom(classPathSupplierClass) && String.class.equals(
+                  ((ParameterizedType) classPathSupplierClass.getGenericSuperclass()).getActualTypeArguments()[0])) {
+
+                classpathSupplier = (Supplier<String>) classPathSupplierClass.newInstance();
+              } else {
+                LOGGER.warn(
+                    "classpath supplier must be subtype of java.util.function.Supplier: " + classpathSupplierClassName);
+              }
+            } catch (ReflectiveOperationException e) {
+              LOGGER.warn("unable to instantiate classpath supplier: " + classpathSupplierClassName, e);
+            }
+            _INSTANCE = new FastSerdeCache(classpathSupplier);
+          } else if (classPath != null) {
+            _INSTANCE = new FastSerdeCache(classPath);
+          } else {
+            // Infer class path if no classpath specified.
+            classPath = System.getProperty("java.class.path");
+            String avroSchemaClassName = "org.apache.avro.Schema";
+            try {
+              Class avroSchemaClass = Class.forName(avroSchemaClassName);
+              String avroLibLocation = avroSchemaClass.getProtectionDomain().getCodeSource().getLocation().getFile();
+              // find all other libs
+              File avroLibFile = new File(avroLibLocation).getParentFile();
+              for (File lib : avroLibFile.listFiles()) {
+                if (lib.getName().endsWith(".jar")) {
+                  classPath += ":" + lib.getAbsolutePath();
+                }
+              }
+            } catch (ClassNotFoundException e) {
+              throw new RuntimeException("Failed to find class: " + avroSchemaClassName);
+            }
+            _INSTANCE = new FastSerdeCache(classPath);
+          }
+        }
+      }
+    }
+    return _INSTANCE;
+  }
+
+  public static boolean isSupportedForFastDeserializer(Schema.Type readerSchemaType) {
+    return readerSchemaType.equals(Schema.Type.RECORD) || readerSchemaType.equals(Schema.Type.MAP)
+        || readerSchemaType.equals(Schema.Type.ARRAY);
+  }
+
+  public static boolean isSupportedForFastSerializer(Schema.Type schemaType) {
+    return schemaType.equals(Schema.Type.RECORD) || schemaType.equals(Schema.Type.MAP) || schemaType.equals(
+        Schema.Type.ARRAY);
+  }
+
+  /**
+   * Generates if needed and returns specific-class aware avro {@link FastDeserializer}.
+   *
+   * @param writerSchema
+   *            {@link Schema} of written data
+   * @param readerSchema
+   *            {@link Schema} intended to be used during deserialization
+   * @return specific-class aware avro {@link FastDeserializer}
+   */
+  public FastDeserializer<?> getFastSpecificDeserializer(Schema writerSchema, Schema readerSchema) {
+    String schemaKey = getSchemaKey(writerSchema, readerSchema);
+    FastDeserializer<?> deserializer = fastSpecificRecordDeserializersCache.get(schemaKey);
```

Comment:
I guess we do a `get()` followed by `putIfAbsent()` here instead of `computeIfAbsent()` because of JDK-8161372, right?

Reply:
Probably since this is the original implementation..